.pre-commit-config.yaml

```diff
@@ -0,0 +1,40 @@
+---
+# See https://pre-commit.com for more information
+fail_fast: false
+
+exclude: \.patch$
+
+# See https://pre-commit.com/hooks.html for more hooks
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.2.0
+    hooks:
+      - id: trailing-whitespace
+      - id: end-of-file-fixer
+      - id: detect-aws-credentials
+        args: ["--allow-missing-credentials"]
+      - id: detect-private-key
+
+  - repo: https://github.com/adrienverge/yamllint
+    rev: v1.26.3
+    hooks:
+      - id: yamllint
+
+  - repo: https://github.com/igorshubovych/markdownlint-cli
+    rev: v0.31.1
+    hooks:
+      - id: markdownlint
+
+  - repo: https://github.com/PyCQA/flake8
```

Comment:
`flake8`, `black`, `autopep8` and bunch of other tools have been superseded by [ruff](https://github.com/astral-sh/ruff-pre-commit).

Reply:
Oh yeah, good point

Reply:
Although I left out the formatter because too many things needed changing and it made all the `versions.py` inconsistent.

Maybe ruff is configurable enough that we can let it do formatting.

Reply:
Done in c6f1fd6.

Side-note: I've disabled flake8, pylance, black in vscode, and am using the ruff vscode extension with the (beta) LSP enabled.